console/v1/types_console_sample.go

```diff
@@ -0,0 +1,166 @@
+	//
+	// It is optional and must be no more than 20 characters in length.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MaxLength=20
+	// +optional
+	Provider string `json:"provider"`
+
+	// badge is a short visual annotation to group samples.
+	//
+	// It is optional and must be no more than 20 characters in length.
+	//
+	// Recommendation is a singular term like "Builder Image", "Devfile" or "Serverless Function".
+	//
+	// Currently, the badge is displayed on sample card tile in the top right corner.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MaxLength=20
+	// +optional
+	Badge string `json:"badge"`
+
+	// tags like "Java", "Quarkus", etc.
+	// Tags are optional and can be used by the user to find samples in the samples catalog.
+	// They will be also displayed on the samples details page.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Tags []string `json:"tags"`
+
+	// The actual sample source, like an external git repository or container image.
+	// +kubebuilder:validation:Required
+	// +required
+	Source ConsoleSampleSource `json:"source"`
+}
+
+// ConsoleSampleSourceType is an enumeration of the supported sample types.
+// Unsupported samples types will be ignored in the web console.
+// +kubebuilder:validation:Enum:=GitImport;ContainerImport
+type ConsoleSampleSourceType string
+
+const (
+	// A sample that let the user import code from a git repository.
+	GitImport ConsoleSampleSourceType = "GitImport"
+	// A sample that let the user import a container image.
+	ContainerImport ConsoleSampleSourceType = "ContainerImport"
+)
+
+// ConsoleSampleSource is the actual sample definition and can hold different sample types.
+// Unsupported sample types will be ignored in the web console.
+// +union
+// +kubebuilder:validation:XValidation:rule="self.type == 'GitImport' ? has(self.gitImport) : !has(self.gitImport)",message="source.gitImport is required when source.type is GitImport, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="self.type == 'ContainerImport' ? has(self.containerImport) : !has(self.containerImport)",message="source.containerImport is required when source.type is ContainerImport, and forbidden otherwise"
+type ConsoleSampleSource struct {
+	// type of the sample, currently supported: "GitImport";"ContainerImport"
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:="GitImport";"ContainerImport"
+	// +kubebuilder:validation:Required
+	Type ConsoleSampleSourceType `json:"type"`
+
+	// gitimport let the user import code from a git repository.
+	// +unionMember,optional
+	GitImport *ConsoleGitImportSample `json:"gitImport,omitempty"`
+
+	// containerimport let the user import a container image.
+	// +unionMember,optional
+	ContainerImport *ConsoleContainerImportSample `json:"containerImport,omitempty"`
+}
+
+// ConsoleGitImportSample let the user import code from a public git repository.
+type ConsoleGitImportSample struct {
+	// Git repository URL.
```

Comment:
For other URLs we use a kubebuilder pattern to check it starts with `https://`, can we add here too?

Reply:
Because Git URLs can also start with ssh:// and git://. I would avoid this. I checked the BuildConfig CRD and the Shipwright.io Build CRD, and both didn't specify a pattern for their URL property. Let me know if you still think it's required.

See also
* https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols
* https://stackoverflow.com/questions/31801271/what-are-the-supported-git-url-formats

Reply:
Yeah that's a fair argument, I'd like to see some examples of the different types of URLs within the godoc please so that others don't ask this question again